Add function to Optionally get `site.info` if not present

Sweetdevil144 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
<!--- Describe your changes in detail -->
This PR aims to refactor site.id and improve str_ns. The end goal is to make the the System independent of DB. Currently, I'm refactoring to create a siteID if not present already. I'll add test cases to check this util function too.
Work still in pending

Some comments from @mdietze :
> 1. Here's my slightly different interpretation :
> 2. Require the user to input a dataframe with lat/lon and optionally siteID; if siteID is not provided, >construct a unique identifier from lat/lon
> 3. replace str_ns with provided unique identifier
> 4. Provide a helper function that takes in a BETY connection and site IDs and returns lat, lon, and str_ns

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
May Fix a Subtask of #3307 
## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [ ] Within one week
- [x] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

Sweetdevil144 committed:
Accept site.data as function input

In do_conversion.R

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Pass site.data via args

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Read Command Line arguments

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update str_ns declaration

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update site$lon

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update if-else block for site.id

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Add helper function to get new.site

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Roxygenise the documentations

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update Function Name

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 commented:
Seems the tests fail as the package is not published with with my Updated function. Should I utilize `source` to Update Imports and later on turn it to `PEcAn.remote::` before it's published?

Sweetdevil144 committed:
Export dependency to DB rather than utils

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update test case

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update 'exampes'

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 commented:
@meetagrawal09 can you cross check if corresponding changes in `test.met.process` are valid or not? :)

Sweetdevil144 committed:
Return null when site$id is not present

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Remove NULL calling

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Fixing 'no global binding err'

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
update global variable handling

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Remove global binding changes

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update global variables

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Ipdate all instances of latlon

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Roxygenise documents

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Usage Updates added

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Doc Updates added

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

meetagrawal09 commented on modules/data.atmosphere/tests/testthat/test.met.process.R:
```diff
@@ -8,7 +8,7 @@ test_that("`met.process` able to call .download.raw.met.module based on met proc
   mockery::stub(met.process, 'PEcAn.DB::dbfile.check', list(id = 1))
   mockery::stub(met.process, 'assign', 1)
-  mockery::stub(met.process, 'PEcAn.DB::query.site', list(lat = 0, lon = 0))
+  mockery::stub(met.process, 'get.new.site', list(new.site = data.frame(id = 1, lat = 0, lon = 0), str_ns = "0-0"))
```
Update this to `PEcAn.DB::get.new.site` rather than just `get.new.site`.

infotroph commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,53 @@
+# http://opensource.ncsa.illinois.edu/license.html
+#-------------------------------------------------------------------------------
+
+##' Get new site info using provided site information
```
Can you explain why this function is needed? Using the database to generate new IDs isn't going to help with the goal of enabling runs of `do_conversions` when no database is present at all.

Sweetdevil144 replied:
This version of file handles only when con != NULL.
I've updated the code to handle db-less ID handling. Can you check it out once?

infotroph commented on base/workflow/R/do_conversions.R:
```diff
@@ -9,6 +9,10 @@
+
+  # extract site info from settings to be utilizes later in processing.
+  site.info <- settings$run$site
+
```
Better to use it directly from inside the existing `settings` object

Sweetdevil144 replied:
I'd been experimenting to see if we have any instance which can be directly utilized with `site` rather than querying it. Although now I've shifted it to a new function, it's nice to use it from within. Thanks for suggestion

infotroph commented:
I think you're taking the wrong approach to task 2 here: Yes, "if siteID is not provided, construct a unique identifier from lat/lon", but it needs to be constructed without using the DB. As we discussed in Slack, this could be as simple as something like `id_str <- paste0(lat, "_", lon)`.

Sweetdevil144 committed:
Update test files

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

Sweetdevil144 committed:
Refactor get.new.site function to generate site IDs and handle missing values when con is not provided

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update site utilizations

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update documentations

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Refactor Errors in check

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update conn pass

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 commented:
Current code is updated for db-less interaction. What more improvements can be made? Would creating a new site-id be correct for a pre existing `lat` and `lon` or should we consider fetching it from `db` if connection is present. Else-if `con=NULL`, we can generate a new siteID. We can create a `int+str` (For eg : `dcf544e6c` etc) styled site-id to distinguish this `str_ns` from those `site-id`'s when `con!=NULL`.

Sweetdevil144 committed:
Update ic_process

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Update Logger statements

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 commented:
Should we add a unit test to check if `get.new.site()` works correctly?

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

infotroph commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,153 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2012 University of Illinois, NCSA.
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the
+# University of Illinois/NCSA Open Source License
+# which accompanies this distribution, and is available at
+# http://opensource.ncsa.illinois.edu/license.html
+#-------------------------------------------------------------------------------
```
```suggestion
```

Sweetdevil144 committed:
Update base/db/R/get.new.site.R

Co-authored-by: Chris Black <chris@ckblack.org>

infotroph commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,153 @@
+
+        if (is.null(site$id) | is.na(site$id)) {
+            if ((!is.null(site$lat) && !is.null(site$lon)) |
+                (!is.na(site$lat) && !is.na(site$lon))
```
Weird little gotcha here -- the !is.null will fire when lat and lon are NA, and the !is.na will return a missing value when lat and lon are null.

infotroph replied:
You probably want `&&` instead of `|`

infotroph commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,145 @@
+                    lat = site$lat,
+                    lon = site$lon
+                )
+                str_ns <- paste0(new.site$lat, "-", new.site$lon)
```
Might want to use a string other than "-" to concatenate here, to avoid confusion between separator characters and negative lon values

Sweetdevil144 replied:
Replaced current logic with a "_" sign

infotroph commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    site.info <- list(new.site = new.site, str_ns = str_ns)
```
Q in live review: why this structure? A: to match what's expected at met.process line 165

Sweetdevil144 replied:
Detailed Explanation : Current structure is followed to match data flow of what information is later on utilised in `met.process` for other function calls

infotroph commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,145 @@
+}
+
+# Function to generate a new siteID (db-less runs ONLY)
+generate_new_siteID <- function() {
+    # Generate a random number. Assuming higher order integers to increase randomness in IDs
+    random_id <- sample(10000:99999999, 1)
+    return(as.numeric(random_id))
+}
```
We want IDs to be deterministic -- don't need this

Sweetdevil144 replied:
Can you suggest a method to perform so? I am unable to determine one right now :) !!

Sweetdevil144 replied:
One approach I suggest is as follow : Hashing to generate a unique id using lat and lon of the site. This would ensure greater precision upto 8 decimal points.

dlebauer replied:
I've implemented hasing in #3423 thus:

```r
mutate(id = digest::digest(geometry, algo = "xxhash64"))
```

where 'geometry' could be replaced by `paste0(lat, lon)` or, if we wanted to be consistent and convert it to an sf object (I'm not certain I see the point) but could do 

```r
data.frame(lat=100, lon = -90) |> 
  sf::st_as_sf(coords = c("lon", "lat"), crs = 4326) |> 
  digest::digest(algo="xxhash64")
```

infotroph commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,145 @@
+
+    # Return the siteID
+    return(siteID)
+}
```
replace with paste0(lat,lon) or slight variant -- it's OK (at least in the met processing workflow) for IDs to be strings that contain minus signs.

infotroph commented on base/workflow/R/do_conversions.R:
```diff
@@ -52,7 +53,7 @@ do_conversions <- function(settings, overwrite.met = FALSE, overwrite.fia = FALS
     # <useic>TRUE</useic>
     if (ic.flag) {
-      settings <- PEcAn.data.land::ic_process(settings, input, dir = dbfiles, overwrite  = overwrite.ic)
+      settings <- PEcAn.data.land::ic_process(settings, input, dir = dbfiles, overwrite  = overwrite.ic, site = settings$run$site)
```
Since site is already in settings, no need to pass as a new argument

infotroph commented on modules/data.atmosphere/R/met.process.R:
```diff
@@ -161,11 +161,12 @@ met.process <- function(site, input_met, start_date, end_date, model,
+  site.info <- PEcAn.DB::get.new.site(site, con=con, latlon = latlon)
+
+  # extract new.site and str_ns from site.info
+  new.site <- site.info$new.site
```
Hypothesis to confirm before acting: No downstream object uses new.site$id, can remove it here

Sweetdevil144 replied:
`new.site$id` is later on passed on to `download.raw.met.module` for further conversions in `convert_input` function. I guess for that instance we will have to keep it rather than duplicating the code.

infotroph commented on modules/data.land/R/ic_process.R:
```diff
@@ -6,14 +6,14 @@
 ##'
 ##' @author Istem Fer, Hamze Dokoohaki
-ic_process <- function(settings, input, dir, overwrite = FALSE){
+ic_process <- function(settings, input, dir, overwrite = FALSE, site = settings$run$site){
```
don't need this new argument, can just keep line 16 below

infotroph commented on modules/data.land/R/soil_process.R:
```diff
@@ -28,13 +28,14 @@ soil_process <- function(settings, input, dbfiles, overwrite = FALSE,run.local=T
+
+  # extract new.site and str_ns from site.info
+  new.site <- site.info$new.site
+  str_ns <- site.info$str_ns
```
I need to think more, but I think we could leave this as it is to be refactored separately -- soil_process already assumes DB is available for other reasons

Sweetdevil144 replied:
Refactoring the same!!

infotroph reviewed (changes requested):
Posting my notes from our live review -- they're a bit cryptic, but hopefully enough to remind you of the key points from the conversation. 

Basically I think the entire get.new.site function is trying to handle too many cases, and instead when the db isn't available we should simply paste together lat and lon to use as a siteid.

Sweetdevil144 committed:
remove browndog support from benchmark

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

infotroph commented on modules/benchmark/DESCRIPTION:
```diff
@@ -40,8 +40,7 @@ Imports:
     PEcAn.data.land,
-    testthat (>= 2.0.0),
-    BrownDog
+    testthat (>= 2.0.0)
```
This should be in #3348 instead of this branch

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Refactor changes from review

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Implement temporary hashing to generate UID

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Remove unrequired changes which were accidently pushed to this branch

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 reviewed (commented):
From previous comments on #3348, I couldn't resist removing unnecessary changes. I will post a discussion regarding this PR and gather insights from other 'members & contributors' too.

Sweetdevil144 committed:
Refactor Dependency chart

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Using openssl for hashing lat/lon

Signed-off-by: Abhinav Pandey <abhinav.pandey.met22@itbhu.ac.in>

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'develop' into siteID-refactor

mdietze commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,123 @@
+##' @title Get New Site Info
+##' @param site a dataframe with site information including id, lat, lon, and time_zone.
+##' @param con Database connection object.
+##' @param latlon Optional global latlon object which will store updated lat/lon.
```
I'm not following why latlon exists as an option here, as information is never extracted from it in any of the cases below within this function, and none of the calls you implemented to this function pass in anything other than NULL.

Sweetdevil144 replied:
`latlon` may exist within `site` as `site$lat` and `site$lon`. In case user had previously extracted this data within `latlon`, then passing `latlon` would seem like an ideal approach. Else the functions extracts required site information from `site` info itself.

Should I delete the `latlon` variable and try to extract information from `site` info itself everytime?

dlebauer replied:
```suggestion
```

dlebauer replied:
Removed `latlon` because it is never used in the function. It was created later whether or not it existed, but not returned. https://github.com/PecanProject/pecan/pull/3324/commits/8132ae69e4d40482c969e1e5f6e07ce9371ed13e

mdietze commented on modules/data.atmosphere/R/met.process.R:
```diff
@@ -133,11 +133,12 @@ met.process <- function(site, input_met, start_date, end_date, model,
-                         lat = latlon$lat, 
-                         lon = latlon$lon)
-  str_ns <- paste0(new.site$id %/% 1e+09, "-", new.site$id %% 1e+09)
+  latlon <- NULL
```
Why is latlon defined to be null here? Why not just delete this line and the `, latlon = latlon` in the function call as you already set the default for latlon to NULL.

mdietze commented on modules/data.land/R/ic_process.R:
```diff
@@ -49,15 +50,25 @@ ic_process <- function(settings, input, dir, overwrite = FALSE){
-                         lon = latlon$lon)
+  latlon <- NULL
+  if(is.null(site$lat) | is.null(site$lon)) {
+    site.info <- PEcAn.DB::get.new.site(site, con=con, latlon = latlon)
```
Same point as earlier about latlon not needing to be defined here.

mdietze commented on modules/data.land/R/ic_process.R:
```diff
@@ -49,15 +50,25 @@ ic_process <- function(settings, input, dir, overwrite = FALSE){
+      id = as.numeric(site$id),
+      lat = site$lat,
+      lon = site$lon
+    )
```
I'm not following why this if/else exists. Doesn't get.new.site already support this `else` case? And if not, shouldn't it?

mdietze commented on modules/data.land/R/soil_process.R:
```diff
@@ -30,6 +30,7 @@ soil_process <- function(settings, input, dbfiles, overwrite = FALSE,run.local=T
   on.exit(PEcAn.DB::db.close(con), add = TRUE)
+  
   # get site info
   latlon <- PEcAn.DB::query.site(site$id, con = con)[c("lat", "lon")]
```
Not following why this function only has whitespace change as it seems like the code chunk here is doing the same thing as what you were trying to support with your new function, and also would benifit from having a database-independent option.

Sweetdevil144 committed:
Merge branch 'develop' of https://github.com/Sweetdevil144/pecan into siteID-refactor

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

dlebauer commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,123 @@
+
+
+# Function to generate a new siteID using hashing (db-less runs ONLY)
+generate_new_siteID <- function(lat, lon) {
```
I think this is a useful function to have - and would like to use it. 

I've implemented something similar as:

```
mutate(id = digest::digest(geometry, algo = "xxhash64"))
```

where 'geometry' could be replaced by paste0(lat, lon) or, if we wanted to be consistent and convert it to an sf object (I'm not certain I see the point) but could do

```
data.frame(lat=lat, lon = lon) |> 
  sf::st_as_sf(coords = c("lon", "lat"), crs = 4326) |> 
  digest::digest(algo="xxhash64")
```

I think both approaches have the same goal - I'd go with whichever one has a lower probability of generating the same id for different locations (unless both are trivially small like <<10^-9). Also, I'd suggest the function be documented, exported, and named generate_site_id (following db field name convention)

Sweetdevil144 replied:
I've documented the function & exported as well. I'd like to go with below functional convention because this seems like the simplest thing to do for now :)
Although if you suggest I can change it to use the latter implementation.
```r
  latlon_str <- paste0(lat, lon)
  uid <- digest::digest(latlon_str, algo = "xxhash64")
```

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

Sweetdevil144 committed:
Merge branch 'PecanProject:develop' into siteID-refactor

Sweetdevil144 committed:
Export generate_new_site, refactor inefficiency while calling get.new.site

Sweetdevil144 committed:
fix R check NOTE

Sweetdevil144 committed:
Update NAMESPACE

Signed-off-by: Abhinav Pandey <abhinavpandey1230@gmail.com>

Sweetdevil144 committed:
Add DESCRIPTION

Signed-off-by: Abhinav Pandey <abhinavpandey1230@gmail.com>

dlebauer committed:
Merge branch 'develop' into siteID-refactor

Sweetdevil144 commented:
Fixing the CI checks and test errors

dlebauer committed:
remove latlon arg from get.new.site

resolving https://github.com/PecanProject/pecan/pull/3324/files/3c537f6552023ceb415ccd139383aefc831f7bca#r1918394219

mdietze commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,121 @@
+##' @return a dataframe with new site information on lat, lon and time_zone
+##' @export
+##' @author Abhinav Pandey
+##' @importFrom digest digest
```
What does this do? First, as far as I can tell this function is never used here. Second, where it is used I don't see why it can't be called by namespace

mdietze commented on base/db/NAMESPACE:
```diff
@@ -63,6 +65,7 @@ export(try2sqlite)
 export(var_names_all)
 export(workflow)
 export(workflows)
+importFrom(digest,digest)
```
Should not be needed

mdietze commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,121 @@
+#' @export
+generate_site_id <- function(lat, lon) {
+  latlon_str <- paste0(lat, lon)
+  uid <- digest::digest(latlon_str, algo = "xxhash64")
```
Do we really need to add an entirely new package dependency to PEcAn just for this one call in one function? Is there an easier solution, or one that relies on existing packages?

dlebauer replied:
Doesn't necessarily warrant using it, but that was my suggestion since I was using it in ccmmf workflows, and it was already a dependency b/c of the api. https://github.com/PecanProject/pecan/blob/7faa280e4eaeecdd2370d3fd5abd14180cc1fafb/apps/api/R/auth.R#L14

infotroph replied:
There's also `rlang::hash`

mdietze commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,121 @@
+            ) {
+                site.id <- paste0(site$lat, "_", site$lon)
+                new.site <- data.frame(
+                    id = as.numeric(site.id),
```
the as.numeric here seems like it's going to always return NA given how site.id is constructed

mdietze commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,121 @@
+                    lat = site$lat,
+                    lon = site$lon
+                )
+                str_ns <- paste0(new.site$lat, "_", new.site$lon)
```
this seems redundant with site.id

mdietze commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,121 @@
+            # Use Pre-existing normalised lat/lon
+            else {
+                PEcAn.logger::logger.warn("Site dataframe does not have an id column")
+                site.id <- generate_site_id()
```
This function has two non-optional arguments that aren't being specified, so how is this code not throwing an error when you run it?

mdietze commented on base/db/R/get.new.site.R:
```diff
@@ -0,0 +1,121 @@
+                PEcAn.logger::logger.info(paste0("Generated siteID using lat and lon:", site.id))
+                site.id <- generate_site_id(site$lat, site$lon)
+                new.site <- data.frame(
+                    id = as.numeric(site.id),
```
does this id need to be numeric? If so, why? Is the algorithm generating it guaranteed to produce numbers?

dlebauer replied:
It won't be numeric if using digest as is.

mdietze commented on modules/data.atmosphere/R/met.process.R:
```diff
@@ -142,11 +142,11 @@ met.process <- function(site, input_met, start_date, end_date, model,
-                         lat = latlon$lat, 
-                         lon = latlon$lon)
-  str_ns <- paste0(new.site$id %/% 1e+09, "-", new.site$id %% 1e+09)
+  site.info <- PEcAn.DB::get.new.site(site, con=con, latlon = latlon)
```
latlon undefined

mdietze commented on modules/data.land/R/ic_process.R:
```diff
@@ -49,21 +50,16 @@ ic_process <- function(settings, input, dir, overwrite = FALSE){
-                         lon = latlon$lon)
+  # Setup site database number, lat, lon and name and copy for format.vars if new input
+  # Then extract new.site and str_ns from site.info 
+  site.info <- PEcAn.DB::get.new.site(site, con=con, latlon = latlon)
```
latlon undefined

mdietze commented on modules/data.land/R/ic_process.R:
```diff
@@ -6,14 +6,15 @@
 ##' @param input Taken from settings$run$inputs. This should include id, path, and source
 ##' @param dir settings$database$dbfiles
 ##' @param overwrite Default = FALSE. whether to force ic_process to proceed
+##' @param site Current site information
```
This line is defining a new argument to the function, but the argument has't been added to the function itself

mdietze commented on base/db/DESCRIPTION:
```diff
@@ -71,7 +72,8 @@ Suggests:
     tidyverse,
-    withr
+    withr,
+    openssl
```
Package added as a dependency, but I can't find any places in your PR where it's actually used. Can this be removed?

mdietze reviewed (changes requested):
What are the examples you are using to test your code as part of development? I ask because (1) there are no testthat tests defined for your new functions and (2) there's a good bit of code here that doesn't look like it would run at all. I want to make sure your not just writing code that "makes sense" and submitting it without verifying that the individual functions  work as expected under all the different cases and that the overall workflow still works

infotroph commented:
@Sweetdevil144 Thank you for the extensive work you put into this PR. It has become quite large compared to the simple idea it started from, and at the same time we've been gaining more experience running PEcAn with no database and learning how much we can simplify things by making more inputs truly required rather than trying to fill them in for you. Specifically, it's now clear that we almost always want to force the user to provide a site id rather than try to create it for them, and that in the cases we create an ID on the fly we should generally be thinking of it as a site name (i.e. something that lets a human figure out which site's in this folder) rather than guarantee its permanence or uniqueness as a global identifier.

I've taken the liberty of cherry-picking commits from this PR into a much smaller set of changes in #3656, which will close this one. I think the replacement PR captures the core ideas of this one (avoiding DB access when we know location already, allowing met download from locations with no preassigned siteid) with a much smaller footprint and better alignment with our current run setup approaches.

Sweetdevil144 commented:
Let's close this Pr once #3656 is merged
